enhancements/microshift/y-minus-2-upgrades.md

```diff
@@ -0,0 +1,270 @@
+
+1. Edge device administrator deploys a host with MicroShift 4.Y-2
+   installed.
+2. Software runs, time passes.
+3. Edge device administrator updates the host to run MicroShift 4.Y.
+  * For ostree-based systems, the host is automatically rebooted as
+    part of the update process.
+  * For RPM-based systems, the user must reboot the host after the
+    software update is completed.
+4. Edge device restarts.
+5. MicroShift restarts.
+6. MicroShift checks the data and binary version difference for
+   compatibility.
+7. If the check fails, MicroShift exits with an error.
+8. If the check passes, MicroShift continues to run, including
+   performing any data migration necessary.
+
+### API Extensions
+
+N/A
+
+### Risks and Mitigations
+
+There is a risk that some underlying data format will change between
+MicroShift versions (kubernetes storage versions, etcd file format,
+etc.). If that happens, someone will have to build a tool to support
+migrating from 4.Y-2 to 4.Y-1 _anyway_. MicroShift will need to carry
+over the use of that tool for an extra release to support the 2
+version upgrade capability.
+
+If we extend the supported upgrade skew, we would have to continue to
+carry the migration tool for the full length of the allowed upgrade
+window after 4.Y-1 (if the allowed skew is 5, we would carry the tool
+in 4.Y-1, 4.Y, 4.Y+1, 4.Y+2, and 4.Y+3 to support upgrading 4.Y-1 to
+4.Y+3 at one time).
+
+The [kubernetes version skew
+policy](https://kubernetes.io/releases/version-skew-policy/) is
+written assuming multi-node clusters. Even so, it supports 3
+kubernetes version difference between the API server and kubelet and 1
+version between the API server instances. This is what allows
+OpenShift's EUS upgrade process, in which the control plane is updated
+independently of the worker nodes, to work. In a single-node
+MicroShift deployment, the API server and kubelet are in the same
+binary and have the same version, so there is no skew at all.
+
+If, in the future, MicroShift does need to support multi-node
+deployments there will be many other aspects of deployment and upgrade
+to consider, in addition to the version skew problem. We can envision
+implementing a process similar to what OpenShift uses, where the
+control plane and workers are updated using separate steps. This would
+make the single-node configuration of MicroShift and the multi-node
+configuration mirror the trade-offs of being able to upgrade the
+entire cluster at one time or offering no downtime that are present in
+SNO and HA OCP.
+
+If an upgrade fails, even after a complex data migration, MicroShift's
+rollback process is to discard the new database and restore the old
+version from a backup before continuing. This ensures that an old
+version of the software matches the older database (file format,
+schema, and content).
+
+MicroShift does not automatically create `StorageVersionMigration` CRs
+to trigger data migration. The core kubernetes APIs are safe because
```

Comment:
While this is true right now, if we were to incorporate something that did require data migrations in the future (e.g. a monitoring stack transitioning from using elasticsearch for log storage to loki), we might need to carry the same version of the monitoring stack through a couple of microshift releases in order to ensure that data migrations take place.

Reply:
Do you mean the migration stack? I don't think monitoring is involved in the storage migration, right? At least not the way we're using it.